subscriptions/subscriptions-impl/src/main/java/com/duckduckgo/subscriptions/impl/ui/SubscriptionWebViewViewModel.kt

```
@@ -0,0 +1,185 @@
+    private val jsonAdapter: JsonAdapter<SubscriptionOptionsJson> = moshi.adapter(SubscriptionOptionsJson::class.java)
+
+    private val command = Channel<Command>(1, DROP_OLDEST)
+    internal fun commands(): Flow<Command> = command.receiveAsFlow()
+
+    data class CurrentPurchaseViewState(val purchaseState: PurchaseStateView = Inactive)
+
+    private val _currentPurchaseViewState = MutableStateFlow(CurrentPurchaseViewState())
+    val currentPurchaseViewState = _currentPurchaseViewState.asStateFlow()
+
+    fun start() {
+        viewModelScope.launch {
+            subscriptionsManager.currentPurchaseState.collect {
+                val state = when (it) {
+                    is CurrentPurchase.Failure -> Failure(it.message)
+                    is CurrentPurchase.Success -> Success
+                    is CurrentPurchase.InProgress, CurrentPurchase.PreFlowInProgress -> InProgress
+                    is CurrentPurchase.Recovered -> Recovered
+                    is CurrentPurchase.PreFlowFinished -> Inactive
+                }
+                _currentPurchaseViewState.emit(currentPurchaseViewState.value.copy(purchaseState = state))
+            }
+        }
+    }
+
+    fun processJsCallbackMessage(featureName: String, method: String, id: String, data: JSONObject) {
+        when (method) {
+            "backToSettings" -> backToSettings()
+            "getSubscriptionOptions" -> getSubscriptionOptions(featureName, method, id)
+            "subscriptionSelected" -> subscriptionSelected(data)
+            else -> {
+                // NOOP
+            }
+        }
+    }
+
+    private fun subscriptionSelected(data: JSONObject?) {
+        viewModelScope.launch(dispatcherProvider.io()) {
+            val id = runCatching { data?.getString("id") }.getOrNull()
+            if (id.isNullOrBlank()) {
+                _currentPurchaseViewState.emit(currentPurchaseViewState.value.copy(purchaseState = Failure("context")))
```

Ignore the error message